### ISSUE TITLE
TimeoutException on wait.until(EC.element_to_be_clickable((By.ID, "faresOutbound"))

### ISSUE THREAD

**Issue by AndrewJDR** [wcrasta/SWA-Scraper#30]
Hello, cool idea for a project!

I gave this a try by running with one of the example commands you provide, but got the following error:
```
$ python3 app.py --company Southwest --depart HOU --arrive MDW --departure-date 05/12/2018 --return-date 05/14/2018 --passengers 2 --max-price 215 --interval 30             
[2018-04-08 22:29:06] Submitting form...
Traceback (most recent call last):
  File "app.py", line 17, in <module>
    main()
  File "app.py", line 12, in main
    real_total = swaScraper.scrape(args)
  File "/home/dev/Documents/src/SWA-Scraper/swaScraper.py", line 78, in scrape
    wait.until(EC.element_to_be_clickable((By.ID, "faresOutbound")))
  File "/usr/local/lib/python3.4/dist-packages/selenium/webdriver/support/wait.py", line 80, in until
    raise TimeoutException(message, screen, stacktrace)
selenium.common.exceptions.TimeoutException: Message: 
Screenshot: available via screen
```
(I actually tried various things to get it to output a screenshot, but couldn't get it to do so -- selenium/phantom is not something I'm terribly familiar with)

Here's my phantomjs version:
```
$ /usr/local/bin/phantomjs --version
2.1.1
```

If you have any ideas, please let me know! thanks.

**Comment by AndrewJDR** [comment]
Also, I'm on linux, and I'm using the 64-bit version of phantomjs.

**Comment by wcrasta** [comment]
Hi @AndrewJDR,

Thanks for your message. Looking at the error log, I can't immediately see what's wrong. When I have some free time this week (probably on the weekend), I'll investigate this issue and get back to you. Wanted to once again thank you for showing interest in the project!

**Comment by wcrasta** [comment]
I'm getting the same error, so it's not just you. My initial guess is that Southwest has changed their website a little bit, making it harder to scrape. I'll do some investigating and see if I can fix the problem.

**Comment by wcrasta** [comment]
After some debugging, I figured out the reason why this is happening. Southwest has completely changed the page that contains flight prices.

Old page: [URL](https://www.southwest.com/flight/select-flight.html?displayOnly=&int=). [Screenshot](https://i.imgur.com/qcsiOlH.png).
New page: [URL](https://www.southwest.com/air/booking/select.html?originationAirportCode=HOU&destinationAirportCode=MDW&returnAirportCode=&departureDate=2018-05-12&departureTimeOfDay=ALL_DAY&returnDate=2018-05-14&returnTimeOfDay=ALL_DAY&adultPassengersCount=2&seniorPassengersCount=0&fareType=USD&passengerType=ADULT&tripType=roundtrip&promoCode=&reset=true&redirectToVision=true&leapfrogRequest=true). [Screenshot](https://i.imgur.com/NgjTA5t.png).

The table with ID "faresOutbound" does not exist in the new page; consequently, the program times out. It should be fairly straightforward to fix. The logic should remain the same, but the names of the HTML elements have changed.

**Comment by wcrasta** [comment]
There are a few reasons why I won't fix this issue right away:

1. I'm quite busy these days and don't have the time to make this fix.
2. Southwest could change their flight prices page again, and all our work will be for naught.

I'll keep this on my TO DO list, albeit it at the bottom. If I have some free time, I'll look into fixing this problem, though I'm not sure when it will be done. **Meanwhile, if somebody else could attempt to fix this issue, I'm sure we would all appreciate it.** Thanks.

**Comment by wcrasta** [comment]
I have some free time today, so I'll try and fix this. Will keep everyone posted.

**Comment by wcrasta** [comment]
Fixed. Let me know if it works. Thanks for alerting me of this problem!

**Comment by AndrewJDR** [comment]
Hey thanks for taking a look!

Unfortunately, I gave this a try (same command as before), and it's not working for me. Very similar error:
```

Traceback (most recent call last):
  File "app.py", line 17, in <module>
    main()
  File "app.py", line 12, in main
    real_total = swaScraper.scrape(args)
  File "/home/dev/Documents/src/SWA-Scraper/swaScraper.py", line 78, in scrape
    wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "currency_dollars")))
  File "/usr/local/lib/python3.4/dist-packages/selenium/webdriver/support/wait.py", line 80, in until
    raise TimeoutException(message, screen, stacktrace)
selenium.common.exceptions.TimeoutException: Message: 
Screenshot: available via screen
```
I gave a shot at trying to fix it myself, but when I looked at the source of the page I couldn't actually find anything with a css class of "currency_dollars". Hmm, I guess it's possible they changed their site already, if so, that was quick!

**Comment by wcrasta** [comment]
Just tried it again, it's working for me. It's possible that they route people to different pages, though I don't think that's what's going on. Let's try debugging it.

In swaScraper.py, can you replace the following lines (81-83):
` except TimeoutError:
            print("[%s] Results took too long!" % (
                datetime.now().strftime("%Y-%m-%d %H:%M:%S")))`

with:
`        except Exception as e:
            print(browser.current_url)`

GitHub isn't allowing me to format the above code properly, but assuming you are familiar with Python, I'm sure you can indent yourself. **And then can you post that URL that is printed?** Thanks.

**Comment by AndrewJDR** [comment]
Here we are:
```
https://www.southwest.com/air/booking/index.html?originationAirportCode=&destinationAirportCode=&returnAirportCode=&departureDate=2018-05-12&departureTimeOfDay=ALL_DAY&returnDate=2018-05-14&returnTimeOfDay=ALL_DAY&adultPassengersCount=2&seniorPassengersCount=0&fareType=USD&passengerType=ADULT&tripType=roundtrip&promoCode=&reset=true&redirectToVision=true&leapfrogRequest=true&validate=true
```
As you can see, it thinks the departure and arrival haven't been filled out. To dig into this a little more I am outputting a screenshot in this "Submit Form" block (lines 65-69):
```
        # Submit Form                                                                                                                                                                                                                                                                       
        search = browser.find_element_by_id("submitButton")                                                                                                                                                                                                                                 
        print("url: %s" % (browser.current_url))                                                                                                                                                                                                                                            
        browser.get_screenshot_as_file('/tmp/sw.png')                                                                                                                                                                                                                                   
        search.click()                     
```
Here's that screenshot:
![sw](https://user-images.githubusercontent.com/1588857/38787793-0f2333b2-40e5-11e8-8301-c96fff3b3e02.png)
Based upon this, it appears to be having difficulty clearing the popup menus for the departure/arrival fields and possibly the date fields as well. Though, I'm not entirely sure if that's actually what is causing an issue. What kind of result do you see if you output a screenshot in that same place?

I tried some things, like sending the tab key just after line 37, like so:
```
depart_date.send_keys(args.departure_date)
depart_date.send_keys(Keys.TAB)
```
But haven't had much luck with that.


That said, it looks like we can just plug the commandline arguments right into this query string, like so:
```
https://www.southwest.com/air/booking/index.html?originationAirportCode=HOU&destinationAirportCode=MDW&returnAirportCode=&departureDate=2018-05-12&departureTimeOfDay=ALL_DAY&returnDate=2018-05-14&returnTimeOfDay=ALL_DAY&adultPassengersCount=2&seniorPassengersCount=0&fareType=USD&passengerType=ADULT&tripType=roundtrip&promoCode=&reset=true&redirectToVision=true&leapfrogRequest=true&validate=true
```
That way the first step of this whole thing doesn't really have to do simulated button presses and things like that?

**Comment by Kerl1310** [comment]
I think the query string idea holds quite a bit of merit as far as design goes. Means we skip a whole step of the process which would speed the whole process up and means that we can implement the same method for every other website we add.
The backend API for these websites is less likely to be changed than the front end too, which means that less maintenance will be needed long-term.

On 16 Apr 2018 04:00, Andrew Johnson <notifications@github.com> wrote:

Here we are:

https://www.southwest.com/air/booking/index.html?originationAirportCode=&destinationAirportCode=&returnAirportCode=&departureDate=2018-05-12&departureTimeOfDay=ALL_DAY&returnDate=2018-05-14&returnTimeOfDay=ALL_DAY&adultPassengersCount=2&seniorPassengersCount=0&fareType=USD&passengerType=ADULT&tripType=roundtrip&promoCode=&reset=true&redirectToVision=true&leapfrogRequest=true&validate=true


As you can see, it thinks the departure and arrival haven't been filled out. To dig into this a little more I am outputting a screenshot in this "Submit Form" block (lines 65-69):

        # Submit Form
        search = browser.find_element_by_id("submitButton")
        print("url: %s" % (browser.current_url))
        browser.get_screenshot_as_file('/tmp/sw.png')
        search.click()


Here's that screenshot:
[sw]<https://user-images.githubusercontent.com/1588857/38787793-0f2333b2-40e5-11e8-8301-c96fff3b3e02.png>
Based upon this, it appears to be having difficulty clearing the popup menus for the departure/arrival fields and possibly the date fields as well. Though, I'm not entirely sure if that's actually what is causing an issue. What kind of result do you see if you output a screenshot in that same place?

I tried some things, like sending the tab key just after line 37, like so:

depart_date.send_keys(args.departure_date)
depart_date.send_keys(Keys.TAB)


But haven't had much luck with that.

That said, it looks like we can just plug the commandline arguments right into this query string, like so:

https://www.southwest.com/air/booking/index.html?originationAirportCode=HOU&destinationAirportCode=MDW&returnAirportCode=&departureDate=2018-05-12&departureTimeOfDay=ALL_DAY&returnDate=2018-05-14&returnTimeOfDay=ALL_DAY&adultPassengersCount=2&seniorPassengersCount=0&fareType=USD&passengerType=ADULT&tripType=roundtrip&promoCode=&reset=true&redirectToVision=true&leapfrogRequest=true&validate=true


That way the first step of this whole thing doesn't really have to do simulated button presses and things like that?

**Comment by dcm684** [comment]
I just made a pull request hopefully fixing the issue, #32. Using a HTTP GET submit, you can directly request ticket prices without having to simulate clicking on a form.

With the booking URL mentioned in previous posts, "booking/index.html,"  you still have to click submit, there is a slightly different URL, "booking/select.html," with the same form arguments that directly gives those ticket prices.

**Comment by AndrewJDR** [comment]
Hrm... I have similar error here even after applying your patch :/

```
$ python3 app.py --company Southwest --depart HOU --arrive MDW --departure-date 05/12/2018 --return-date 05/14/2018 --passengers 2 --max-price 215 --interval 30
[2018-04-24 21:31:34] Submitting form...
Traceback (most recent call last):
  File "app.py", line 17, in <module>
    main()
  File "app.py", line 12, in main
    real_total = swaScraper.scrape(args)
  File "/home/dev/Documents/src/SWA-Scraper/swaScraper.py", line 150, in scrape
    wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "currency_dollars")))
  File "/usr/local/lib/python3.4/dist-packages/selenium/webdriver/support/wait.py", line 80, in until
    raise TimeoutException(message, screen, stacktrace)
selenium.common.exceptions.TimeoutException: Message: 
Screenshot: available via screen
```

**Comment by AndrewJDR** [comment]
Ok. So I changed my phantomjs to beta 2.5.0 on this page:
https://bitbucket.org/ariya/phantomjs/downloads/
And it seems to be working now!

For whatever reason, the latest phantomjs 2.1.1 available from the website simply does not work for me.

**Comment by wcrasta** [comment]
Great discussion here guys, thanks everyone. I'll review @dcm684's PR soon and discuss the merits of simulating clicking through a form vs. entering arguments via URL and will respond.

@AndrewJDR, glad to hear that things are working for you. To be clear, after changing your version of PhantomJS, is my code working, @dcm684's code working, or both?

**Comment by dcm684** [comment]
@AndrewJR, That's interesting, I'm using PhantomJS 2.1.1 and having been running the script with my changes on 2.1.1 for the past two days without any issue. I just tried it with the arguments that gave you a time out error, and it worked. Lowest outbound $115 and lowest return $117.

I am using Windows if that makes any difference.

**Comment by AndrewJDR** [comment]
@dcm684 I'm using Ubuntu 14.04 here, so perhaps that's a factor.

@wcrasta Using phantomjs 2.5.0 beta, things work as expected when using 540762e + PR #32, but I get the following error when I use 540762e by itself:
```
Traceback (most recent call last):
  File "app.py", line 17, in <module>
    main()
  File "app.py", line 12, in main
    real_total = swaScraper.scrape(args)
  File "/home/dev/Documents/src/SWA-Scraper/swaScraper.py", line 24, in scrape
    depart_airport = browser.find_element_by_id("originAirport_displayed")
  File "/home/dev/Documents/src/SWA-Scraper/env/lib/python3.4/site-packages/selenium/webdriver/remote/webdriver.py", line 269, in find_element_by_id
    return self.find_element(by=By.ID, value=id_)
  File "/home/dev/Documents/src/SWA-Scraper/env/lib/python3.4/site-packages/selenium/webdriver/remote/webdriver.py", line 752, in find_element
    'value': value})['value']
  File "/home/dev/Documents/src/SWA-Scraper/env/lib/python3.4/site-packages/selenium/webdriver/remote/webdriver.py", line 236, in execute
    self.error_handler.check_response(response)
  File "/home/dev/Documents/src/SWA-Scraper/env/lib/python3.4/site-packages/selenium/webdriver/remote/errorhandler.py", line 192, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.NoSuchElementException: Message: {"errorMessage":"Unable to find element with id 'originAirport_displayed'","request":{"headers":{"Accept":"application/json","Accept-Encoding":"identity","Connection":"close","Content-Length":"104","Content-Type":"application/json;charset=UTF-8","Host":"127.0.0.1:49837","User-Agent":"Python-urllib/3.4"},"httpVersion":"1.1","method":"POST","post":"{\"sessionId\": \"e9f8ef60-4935-11e8-a619-975a50cd99a7\", \"using\": \"id\", \"value\": \"originAirport_displayed\"}","url":"/element","urlParsed":{"anchor":"","query":"","file":"element","directory":"/","path":"/element","relative":"/element","port":"","host":"","password":"","user":"","userInfo":"","authority":"","protocol":"","source":"/element","queryKey":{},"chunks":["element"]},"urlOriginal":"/session/e9f8ef60-4935-11e8-a619-975a50cd99a7/element"}}
Screenshot: available via screen
```

This is using the command:
```
python3 app.py --company Southwest --depart HOU --arrive MDW --departure-date 05/12/2018 --return-date 05/14/2018 --passengers 2 --max-price 300 --interval 30
```

**Comment by var-che** [comment]
@AndrewJDR I tried the same command as you, and I got the same identical error as you. I believe that the element '_originAirport_displayed_' doesn't exist on that website anymore so the script needs to change. Well, at least that element with that id. Does anyone know what was that element, and what is the current name of it?

OS: Windows 10
phantomjs: 2.1.1-windows

**Comment by AndrewJDR** [comment]
@var-che please try using PR #32

**Comment by var-che** [comment]
@AndrewJDR Thank you for your fast response. It did fix the problem for me, and I recommend that OP consider reviewing pull request.

**Comment by wcrasta** [comment]
My apologies all, I haven't had much time to review the PR. I'm not sure when I'll get to it, but hopefully within the next couple of days. I've been putting it off, since the code change is probably a whole different idea/design and will need quite a bit of reviewing. In the meantime, feel free to add your review/comments to PR #32 to help make my job a little easier. Thanks!

**Comment by eskemojoe007** [comment]
I had my own version of something similar and ran into the same issue.  I used to just be able to use `requests` and not need selenium at all.  I had a function that looked like:

```python
def enter_single_search(origin,destination,outbound_date_str,return_date_str,
    method='dollars',driver=None,
    url='https://www.southwest.com/flight/search-flight.html?',
    available_codes=None, wait_time=.2,
    proxies=None):

    if not driver is None:
        logging.warning('you specified a driver...we no longer use selenium, Driver should always be None...continueing')

    payload = {
        'twoWayTrip': True,
        'airTranRedirect': "",
        'returnAirport': "RoundTrip",
        'outboundTimeOfDay': "ANYTIME",
        'returnTimeOfDay': "ANYTIME",
        'seniorPassengerCount': 0,
        'fareType': "DOLLARS",
        'originAirport': origin,
        'destinationAirport': destination,
        'outboundDateString': outbound_date_str,
        'returnDateString': return_date_str,
        'adultPassengerCount': 1}

    verify = get_verify(proxies)
    r =requests.post(url,params=payload,proxies=proxies,verify=verify)

    logging.info('Got payload back from SW')
    return r
```
I've attempted to do the same with the new URL that the new website gives similar to PR #32, but it gives an empty results.  Just thought I'd share.

**Comment by wcrasta** [comment]
I've merged in PR #32 which hopefully fixes most of the issues. There are some bugs/errors here and there, which I'll need to review and fix. Due to time constraints, I'm not sure when I'll get around to validating that everything works. I would, however, be very grateful to anyone who fixes whatever's broken. Thanks.

**Comment by wcrasta** [comment]
Finally got around to reviewing the code and testing it quite a bit to make sure that it's error-free. Commit 59d32929fbf81461c958156e77e87700de2c0741 should do the job. Thanks especially to @dcm684, who got the ball rolling with the re-write of the code. Going forward, I'll try and do a better job of reviewing/merging code in a more timely manner.
